# nf-core/genomeskim: Output

## Introduction

This document describes the output produced by the pipeline. Most of the plots are taken from the MultiQC report, which summarises results at the end of the pipeline.

The directories listed below will be created in the results directory after the pipeline has finished. All paths are relative to the top-level results directory.

<!-- TODO nf-core: Write this documentation describing your workflow's output -->

## Directory structure

The directory structure after a completed run will look something like the below. Directories surrounded by asterisks like `*gofetch*/` are not created by default and will only be written if the respective argument is used (e.g. `--gofetch_save_outputs` in this case). Other directories are created by default, unless that part of the pipeline is skipped - for example, the `validation/` directory won't be created if `--skip_validation` is used.

```
{outdir}
├── multiqc/
├── pipeline_info/
├── getorganelle/
│   └── <sample>/
├── annotation/
│   └── <sample>/
├── validation/
│   └── <sample>_summary.tsv
├── genomestats/
|   └── <sample>*
└── *gofetch*/
    └── <taxon>/


work/
.nextflow.log
```
There will be a separate `<sample>` file or directory for each sample input in your samplesheet.

More details about each of the outputs can be seen in the below sections:


- [MultiQC](#multiqc) - Aggregate report describing results and QC from the whole pipeline
- [Pipeline information](#pipeline-information) - Report metrics generated during the workflow execution
- [GetOrganelle](#getorganelle) - Assembled organelle contigs
- [Annotation](#annotation) - Annotations and extracted gene sequences
- [Validation](#validation) - Compiled taxonomic and coverage for each set of contigs from each sample
- [Genomestats](#genomestats) - Outputs from genomescope2 estimating genome parameters
- [*GoFetch*](#gofetch) - Downloaded taxonomic references


### MultiQC

<details markdown="1">
<summary>Output files</summary>

- `multiqc/`
  - `multiqc_report.html`: a standalone HTML file that can be viewed in your web browser.
  - `multiqc_data/`: directory containing parsed statistics from the different tools used in the pipeline.
  - `multiqc_plots/`: directory containing static images from the report in various formats.

</details>

[MultiQC](http://multiqc.info) is a visualization tool that generates a single HTML report summarising all samples in your project. Most of the pipeline QC results are visualised in the report and further statistics are available in the report data directory.

Results generated by MultiQC collate pipeline QC from supported tools e.g. FastQC (see below). The pipeline has special steps which also allow the software versions to be reported in the MultiQC output for future traceability. For more information about how to use MultiQC reports, see <http://multiqc.info>.

[FastQC](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/) gives general quality metrics about your sequenced reads. It provides information about the quality score distribution across your reads, per base sequence content (%A/T/G/C), adapter contamination and overrepresented sequences. For further reading and documentation see the [FastQC help pages](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/).

![MultiQC - FastQC sequence counts plot](images/mqc_fastqc_counts.png)

![MultiQC - FastQC mean quality scores plot](images/mqc_fastqc_quality.png)

![MultiQC - FastQC adapter content plot](images/mqc_fastqc_adapter.png)

:::note
The FastQC plots displayed in the MultiQC report shows _untrimmed_ reads. They may contain adapter sequence and potentially regions with low quality.
:::

### Pipeline information

<details markdown="1">
<summary>Output files for all samples</summary>

- `pipeline_info/`
  - Reports generated by Nextflow: `execution_report.html`, `execution_timeline.html`, `execution_trace.txt` and `pipeline_dag.dot`/`pipeline_dag.svg`.
  - Reports generated by the pipeline: `pipeline_report.html`, `pipeline_report.txt` and `software_versions.yml`. The `pipeline_report*` files will only be present if the `--email` / `--email_on_fail` parameters are used when running the pipeline.
  - Reformatted samplesheet files used as input to the pipeline: `samplesheet.valid.csv`.
  - Parameters used by the pipeline run: `params.json`.

</details>

[Nextflow](https://www.nextflow.io/docs/latest/tracing.html) provides excellent functionality for generating various reports relevant to the running and execution of the pipeline. This will allow you to troubleshoot errors with the running of the pipeline, and also provide you with other information such as launch commands, run times and resource usage.

## GetOrganelle

<details markdown="1">
<summary>Output files</summary>

- `getorganelle/<sample>`
  - `*path_sequence.fasta.gz` - assembled contigs
  - `*selected_graph.gfa*` - the final graph for the output assembled contigs
  - `extended_*_paired.fq.gz`, `extended_*_unpaired.fq.gz` - reads used in assembly
  - `*assembly_graph.fastq*` - all assembly graph files
  - `get_org.log.txt` - GetOrganelle's log file
  - `extended_spades/`
    - Files related to running SPAdes within GetOrganelle

</details>

GetOrganelle outputs can vary substantially, the above shows a typical successful output. For more information and to debug tricky assemblies, see GetOrganelle's wiki [here](https://github.com/Kinggerm/GetOrganelle/wiki/FAQ#output)

## Annotation

<details markdown="1">
<summary>Output files</summary>

- `annotation/`
  - `<sample>.fas` - nucleotide sequences of individual genes annotated to contigs for this sample
  - `<sample>.faa` - amino acide sequences of individual genes annotated to contigs for this sample
  - `<sample>.gff`, - `<sample>.bed` - Files describing locations and attributes of annotations on contigs for this sample

</details>

These files detail the annotations found by MITOS on the contigs for each sample, and extract out the individual gene sequences and translations.

## Validation

<details markdown="1">
<summary>Output files</summary>

- `validation/`
  - `<sample>_summary.tsv` - summary of estimated taxonomy and coverage

</details>

These files are compiled by Blobtoolkit to summarise the findings of the BLAST search for taxonomy and the minimap2 coverage mapping.

## Genomestats

<details markdown="1">
<summary>Output files</summary>

- `genomestats/`
  - `<sample>_*_plot.png` - plots of actual and modelled kmer coverage
  - `<sample>_summary.txt` - estimated genome parameters
  - `<sample>_model.txt` - model output and significance tests

</details>

These files are output by Genomescope2 based on an analysis of the kmer counting performed by Jellyfish

## GoFetch

<!-- TODO -->

